pkg/controllers/statusaggregator/controller.go

```diff
@@ -347,15 +342,57 @@ func (a *StatusAggregator) reconcile(ctx context.Context, key reconcileKey) (sta
 		return worker.StatusError
 	}
 
-	newObj, needUpdate, err := plugin.AggregateStatuses(ctx, sourceObject, fedObject, clusterObjs, clusterObjsUpToDate)
-	if err != nil {
-		return worker.StatusError
+	needUpdate := false
+	for _, plugin := range plugins.DefaultCommonPlugins {
+		newObj, newNeedUpdate, err := plugin.AggregateStatuses(ctx, sourceObject, fedObject, clusterObjs, clusterObjsUpToDate)
+		if err != nil {
+			return worker.StatusError
+		}
+		sourceObject = newObj
+		needUpdate = needUpdate || newNeedUpdate
 	}
 
 	if needUpdate {
-		logger.V(1).Info("Updating status of source object")
+		logger.V(1).Info("Updating metadata of source object")
+		_, err = a.dynamicClient.Resource(ftc.GetSourceTypeGVR()).Namespace(key.namespace).
+			Update(ctx, sourceObject, metav1.UpdateOptions{})
+		if err != nil {
+			if apierrors.IsConflict(err) {
+				return worker.StatusConflict
+			}
+			return worker.StatusError
+		}
+	}
+
+	needUpdate = false
+	plugin := plugins.GetPlugin(key.gvk)
+	if plugin != nil {
+		newObj, newNeedUpdate, err := plugin.AggregateStatuses(ctx, sourceObject, fedObject, clusterObjs, clusterObjsUpToDate)
+		if err != nil {
+			return worker.StatusError
+		}
+		sourceObject = newObj
+		needUpdate = needUpdate || newNeedUpdate
+	} else {
+		logger.V(3).Info("Plugin not found")
+	}
+
+	if needUpdate {
+		latestSourceObj, err := a.dynamicClient.Resource(ftc.GetSourceTypeGVR()).Namespace(key.namespace).
```

Comment:
Why do we need to get a fresh copy of the source object?

Reply:
Because we just update the sourceObj early. So we need to get the latest obj before updateStatus.

Reply:
Understood. Would using `patch` avoid the additional get? I'm slightly worried about the burden on the apiserver.

Reply:
My concern is that users may update the annotation on source object in the same time when statusAggregator updates the anno. Using patch may override the changes users made.